java/rocksjni/rocksjni.cc

```diff
@@ -746,6 +747,282 @@ void Java_org_rocksdb_RocksDB_putDirect(
+
+  const auto status =
+      db->GetEntity(default_read_options, columns_family, key_slice, &columns);
+
+  if (status.ok() && columns.columns().size() > 0) {
+    auto max_items = std::min(static_cast<jsize>(columns.columns().size()),
+                              env->GetArrayLength(jnames));
+
+    auto namesOffset = std::make_unique<jint[]>(max_items);
+    env->GetIntArrayRegion(jNamesOffset, 0, max_items, namesOffset.get());
+    if (env->ExceptionCheck() == JNI_TRUE) {
+      return nullptr;
+    }
+
+    auto namesLen = std::make_unique<jint[]>(max_items);
+    env->GetIntArrayRegion(jNamesLen, 0, max_items, namesLen.get());
+    if (env->ExceptionCheck() == JNI_TRUE) {
+      return nullptr;
+    }
+
+    auto namesRequiredSize = std::make_unique<jint[]>(max_items);
+
+    auto valuesOffset = std::make_unique<jint[]>(max_items);
+    env->GetIntArrayRegion(jValuesOffset, 0, max_items, valuesOffset.get());
+    if (env->ExceptionCheck() == JNI_TRUE) {
+      return nullptr;
+    }
+
+    auto valuesLen = std::make_unique<jint[]>(max_items);
+    env->GetIntArrayRegion(jValuesLen, 0, max_items, valuesLen.get());
+    if (env->ExceptionCheck() == JNI_TRUE) {
+      return nullptr;
+    }
+
+    auto valuesRequiredSize = std::make_unique<jint[]>(max_items);
+
+    for (int i = 0; i < max_items; i++) {
+      auto column = columns.columns()[i];
+
+      auto jname = env->GetObjectArrayElement(jnames, i);
+      if (env->ExceptionCheck()) {
+        return nullptr;
+      }
+      auto _name = env->GetDirectBufferAddress(jname);
+      if (_name == nullptr) {
+        ROCKSDB_NAMESPACE::RocksDBExceptionJni::ThrowNew(
+            env,
+            "Invalid \"name\" argument (argument is not a valid direct "
+            "ByteBuffer)");
+        return nullptr;
+      }
+      auto name = reinterpret_cast<char*>(_name) + namesOffset[i];
+      namesRequiredSize[i] = static_cast<jint>(column.name().size());
+      auto max_data_len = std::min(namesRequiredSize[i], (namesLen[i]));
+      std::memcpy(name, column.name().data(), max_data_len);
+
+      auto j_value = env->GetObjectArrayElement(jvalues, i);
+      if (env->ExceptionCheck()) {
+        return nullptr;
+      }
+      auto _value = env->GetDirectBufferAddress(j_value);
+      if (_value == nullptr) {
+        ROCKSDB_NAMESPACE::RocksDBExceptionJni::ThrowNew(
+            env,
+            "Invalid \"value\" argument (argument is not a valid direct "
+            "ByteBuffer)");
+        return nullptr;
+      }
+
+      auto value = reinterpret_cast<char*>(_value) + valuesOffset[i];
+      valuesRequiredSize[i] = static_cast<jint>(column.value().size());
+      max_data_len = std::min(valuesRequiredSize[i], (valuesLen[i]));
+      std::memcpy(value, column.value().data(), max_data_len);
+    }
+
+    env->SetIntArrayRegion(jnamesRequiredSize, 0, max_items,
+                           namesRequiredSize.get());
+    if (env->ExceptionCheck()) {
+      return nullptr;
```

Comment:
Do we need to free the previous JNI array fetches and local-references here?

Reply:
No we don't need to. See previous comment